x/icacallbacks/module_ibc.go

```diff
@@ -108,12 +113,29 @@ func (im IBCModule) CallRegisteredICACallback(ctx sdk.Context, modulePacket chan
 		return err
 	}
 	// fetch the callback data
-	callback, err := im.keeper.GetICACallbackHandler(module)
+	callbackDataKey := types.CallbackDataKeyFormatter(portID, channelID, modulePacket.Sequence)
+	callbackData, found := im.keeper.GetCallbackData(ctx, callbackDataKey)
+	if !found {
+		errMsg := fmt.Sprintf("callback data not found for portID: %s, channelID: %s, sequence: %d", portID, channelID, modulePacket.Sequence)
+		return sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, errMsg)
+	}
+	// fetch the callback function
+	callbackHandler, err := im.keeper.GetICACallbackHandler(module)
 	if err != nil {
 		return err
 	}
 	// call the callback
-	_ = callback
+	if callbackHandler.HasICACallback(callbackData.CallbackId) {
+		// if acknowledgement is empty, then it is a timeout
+		err := callbackHandler.CallICACallback(ctx, callbackData.CallbackId, modulePacket, acknowledgement, callbackData.CallbackArgs)
+		if err != nil {
+			return err
+		}
+	}
+
+	// remove the callback data
+	// NOTE: Should we remove the callback data here, or above (conditional on HasICACallback == true)?
+	im.keeper.RemoveCallbackData(ctx, callbackDataKey)
```

Comment:
Do we want to hold onto the record in the case of a failure?

Reply:
I _think_ the way this works is that if an ack fails to execute, the ack can be retried, so we do want to hold onto this until a successful ack has processed.